modules/pubsub/src/main/scala/com/snowplowanalytics/snowplow/sources/pubsub/PubsubSource.scala

```diff
@@ -7,325 +7,254 @@
 object PubsubSource {
 
-  private implicit def logger[F[_]: Sync]: SelfAwareStructuredLogger[F] = Slf4jLogger.getLogger[F]
+  private implicit def logger[F[_]: Sync]: Logger[F] = Slf4jLogger.getLogger[F]
 
   def build[F[_]: Async](config: PubsubSourceConfig): F[SourceAndAck[F]] =
-    LowLevelSource.toSourceAndAck(lowLevel(config))
-
-  private type PubSubCheckpointer[F[_]] = Checkpointer[F, Chunk[AckReplyConsumer]]
+    Deferred[F, PubsubCheckpointer.Resources[F]].flatMap { deferred =>
+      LowLevelSource.toSourceAndAck(lowLevel(config, deferred))
+    }
 
-  private def lowLevel[F[_]: Async](config: PubsubSourceConfig): LowLevelSource[F, Chunk[AckReplyConsumer]] =
-    new LowLevelSource[F, Chunk[AckReplyConsumer]] {
-      def checkpointer: PubSubCheckpointer[F] = pubsubCheckpointer
+  private def lowLevel[F[_]: Async](
+    config: PubsubSourceConfig,
+    deferredResources: Deferred[F, PubsubCheckpointer.Resources[F]]
+  ): LowLevelSource[F, Vector[Unique.Token]] =
+    new LowLevelSource[F, Vector[Unique.Token]] {
+      def checkpointer: Checkpointer[F, Vector[Unique.Token]] = new PubsubCheckpointer(config.subscription, deferredResources)
 
-      def stream: Stream[F, Stream[F, LowLevelEvents[Chunk[AckReplyConsumer]]]] =
-        pubsubStream(config)
+      def stream: Stream[F, Stream[F, LowLevelEvents[Vector[Unique.Token]]]] =
+        pubsubStream(config, deferredResources)
 
       def lastLiveness: F[FiniteDuration] =
         Sync[F].realTime
     }
 
-  private def pubsubCheckpointer[F[_]: Async]: PubSubCheckpointer[F] = new PubSubCheckpointer[F] {
-    def combine(x: Chunk[AckReplyConsumer], y: Chunk[AckReplyConsumer]): Chunk[AckReplyConsumer] =
-      Chunk.Queue(x, y)
-
-    val empty: Chunk[AckReplyConsumer] = Chunk.empty
-    def ack(c: Chunk[AckReplyConsumer]): F[Unit] =
-      Sync[F].delay {
-        c.foreach(_.ack())
-      }
-
-    def nack(c: Chunk[AckReplyConsumer]): F[Unit] =
-      Sync[F].delay {
-        c.foreach(_.nack())
-      }
-  }
-
-  private case class SingleMessage(
-    message: ByteBuffer,
-    ackReply: AckReplyConsumer,
-    tstamp: Instant
-  )
+  private def pubsubStream[F[_]: Async](
+    config: PubsubSourceConfig,
+    deferredResources: Deferred[F, PubsubCheckpointer.Resources[F]]
+  ): Stream[F, Stream[F, LowLevelEvents[Vector[Unique.Token]]]] =
+    for {
+      parallelPullCount <- Stream.eval(Sync[F].delay(chooseNumParallelPulls(config)))
+      stub <- Stream.resource(stubResource(config))
+      refStates <- Stream.eval(Ref[F].of(Map.empty[Unique.Token, PubsubBatchState]))
+      _ <- Stream.eval(deferredResources.complete(PubsubCheckpointer.Resources(stub, refStates)))
+    } yield Stream
+      .fixedRateStartImmediately(config.debounceRequests, dampen = true)
+      .parEvalMapUnordered(parallelPullCount)(_ => pullAndManageState(config, stub, refStates))
+      .unNone
+      .repeat
+      .prefetchN(parallelPullCount)
+      .concurrently(extendDeadlines(config, stub, refStates))
```

Comment:
Idea: 

As `extendDeadlines` is some process running in the background.... maybe we can keep it as `F[_]` and run as a resource using:

`Async[F].background(...)`

when we build our resources earlier above? Just an idea to consider

Reply:
Your explanation makes perfect sense and I'm happy with the original solution :)